inf-clojure.el

```diff
@@ -1204,52 +1207,60 @@ prefix argument PROMPT-FOR-SYMBOL, it prompts for a symbol name."
+                 (clojure-find-ns))))
+    (inf-clojure--send-string proc (format (inf-clojure-ns-vars-form proc) ns))))
 
 (defun inf-clojure-set-ns (prompt-for-ns)
   "Set the ns of the inferior Clojure process to NS.
 See variable `inf-clojure-set-ns-form'.  It defaults to the ns of
 the current buffer.  When invoked with a prefix argument
 PROMPT-FOR-NS, it prompts for a namespace name."
   (interactive "P")
-  (let ((ns (if prompt-for-ns
+  (let ((proc (inf-clojure--proc-or-error))
+        (ns (if prompt-for-ns
                 (car (inf-clojure-symprompt "Set ns to" (clojure-find-ns)))
               (clojure-find-ns))))
     (when (or (not ns) (equal ns ""))
       (user-error "No namespace selected"))
-    (inf-clojure--send-string (inf-clojure-proc) (format (inf-clojure-set-ns-form) ns))))
+    (inf-clojure--send-string proc (format (inf-clojure-set-ns-form proc) ns))))
 
 (defun inf-clojure-apropos (var)
   "Send a form to the inferior Clojure to give apropos for VAR.
 See variable `inf-clojure-apropos-form'."
-  (interactive (inf-clojure-symprompt "Var apropos" (inf-clojure-symbol-at-point)))
-  (inf-clojure--send-string (inf-clojure-proc) (format (inf-clojure-apropos-form) var)))
+  (interactive)
+  (let ((proc (inf-clojure--proc-or-error))
+        (expression (or (car (inf-clojure-symprompt "Var apropos" (inf-clojure-symbol-at-point)))
+                        (inf-clojure-symbol-at-point))))
+    (inf-clojure--send-string proc (format (inf-clojure-apropos-form proc) expression))))
 
 (defun inf-clojure-macroexpand (&optional macro-1)
   "Send a form to the inferior Clojure to give apropos for VAR.
 See variable `inf-clojure-macroexpand-form'.
 With a prefix arg MACRO-1 uses `inf-clojure-macroexpand-1-form'."
   (interactive "P")
-  (let ((last-sexp (buffer-substring-no-properties (save-excursion (backward-sexp) (point)) (point))))
+  (let ((proc (inf-clojure--proc-or-error))
+        (last-sexp (buffer-substring-no-properties (save-excursion (backward-sexp) (point)) (point))))
     (inf-clojure--send-string
-     (inf-clojure-proc)
+     proc
      (format (if macro-1
-                 (inf-clojure-macroexpand-1-form)
-               (inf-clojure-macroexpand-form))
+                 (inf-clojure-macroexpand-1-form proc)
+               (inf-clojure-macroexpand-form proc))
              last-sexp))))
 
-
 (defun inf-clojure-proc ()
   "Return the current inferior Clojure process.
 See variable `inf-clojure-buffer'."
-  (let ((proc (get-buffer-process (if (derived-mode-p 'inf-clojure-mode)
-                                      (current-buffer)
-                                    inf-clojure-buffer))))
-    (or proc
-        (error "No Clojure subprocess; see variable `inf-clojure-buffer'"))))
+  (get-buffer-process (if (derived-mode-p 'inf-clojure-mode)
+                          (current-buffer)
+                        inf-clojure-buffer)))
+
+(defun inf-clojure--proc-or-error ()
```

Comment:
I think a more common approach is to have just one function and some param controlling wether to raise errors or return `nil`. That's what we do in CIDER for instance (and I've seen in many other packages).

Reply:
The function won't be pure anymore though

Reply:
In Emacs that's totally fine. :-)